src/main/java/org/generationcp/middleware/dao/dms/ObservationUnitsSearchDao.java

```diff
@@ -965,4 +1002,175 @@ private ObservationUnitRow getObservationUnitRow(final ObservationUnitsSearchDTO
+		final String observationVariableName) {
+		try {
+
+			final String sql = this.getObservationUnitTableAsListOfMapQuery(searchDto, observationVariableName);
+			final SQLQuery query = this.getSession().createSQLQuery(sql);
+
+			for (final String columnName : searchDto.getFilter().getFilterColumns()) {
+				query.addScalar(columnName);
+			}
+
+			this.setParameters(searchDto, query);
+			return this.convertSelectionAndTraitColumnsValueType(query.list(), searchDto.getSelectionMethodsAndTraits());
+
+		} catch (final Exception e) {
+			final String error = "An internal error has ocurred when trying to execute the operation " + e.getMessage();
+			ObservationUnitsSearchDao.LOG.error(error);
+			throw new MiddlewareException(error, e);
+		}
+	}
+
+	protected List<Map<String, Object>> convertSelectionAndTraitColumnsValueType(final List<Map<String, Object>> result,
+		final List<MeasurementVariableDto> selectionAndTraits) {
+		final Iterator<Map<String, Object>> iterator = result.iterator();
+		while (iterator.hasNext()) {
+			final Map<String, Object> rowMap = iterator.next();
+			for (final MeasurementVariableDto measurementVariableDto : selectionAndTraits) {
+				// Unfortunately, since the trait values (Numerical or Categorical with numeric code) are stored and returned as String from the database,
+				// We have to manually convert them to numeric data type (if possible) so that the data returned to the client will be properly processed
+				// when we send it to the OpenCPU API. Also, if a trait's value is "missing", we should return it as a null value in order for
+				// OpenCPU to recognize it as 'NA' (Not Available).
+				if (rowMap.containsKey(measurementVariableDto.getName())) {
+					final String stringValue = String.valueOf(rowMap.get(measurementVariableDto.getName()));
+					if (Phenotype.MISSING_VALUE.equals(stringValue)) {
+						rowMap.put(measurementVariableDto.getName(), null);
+					} else {
+						// Convert numeric sting to BigDecimal to support signed decimal number
+						rowMap.put(measurementVariableDto.getName(),
+							NumberUtils.isNumber(stringValue) ? NumberUtils.createBigDecimal(stringValue) :
+								rowMap.get(measurementVariableDto.getName()));
+					}
+				}
+			}
+		}
+		return result;
+	}
+
+	/**
+	 * Stripped-down version of {@Link ObservationUnitsSearchDao#getObservationUnitTableQuery}. It will only return the values (no metadata -- PhenotypeId, experimentId, obs_unit_id, etc) from factors and variates at Observation level.
+	 * The query is also optimized to exclude columns that are not specified in {@Link ObservationUnitsSearchDTO#filter#filterColumns} parameter as to reduce the number of subqueries and ensure that the
+	 * query only returns columns that are needed per request.
+	 *
+	 * @param searchDto
+	 * @param observationUnitNoName
+	 * @return
+	 */
+	private String getObservationUnitTableAsListOfMapQuery(
```

Comment:
As discussed: 
-it's essential for performance in the visualization option to optimize the query to only get information it needs (reduce subqueries etc)
-in the long run, it would be ideal to only maintain one query

Reply:
We can maybe follow this logic wherever you are checking `filterColumns`:
`final boolean noFilterVariables = CollectionUtils.isEmpty(filterColumns);
		if (noFilterVariables || filterColumns.contains(TRIAL_INSTANCE)) {
			sqlColumns.add("    gl.description AS TRIAL_INSTANCE");
		}`

Just make sure to put a comment when filterColumns has a value and when it doesn't

Reply:
@abatac 
I have just one question here: Did you experienced real performance degradation here with the all-columns query or was this decision only based on theoretical information? Does this new query improve the performance of the visualization significantly?

Thanks!

cc @darla-leafnode @clarysabel

Reply:
Hi @nahuel-soldevilla, Yes, I did experience performance degradation using the all-columns query. Tested on a study with 30,000 records, when using the all-columns query, the generation of visualization graph takes around 2 min 30 seconds. When using the new query to return only the columns the visualization graph needs, it only takes around 20 seconds. 

cc @darla-leafnode @clarysabel